### PR TITLE
Kill query blocked in lock tree

### PR DESCRIPTION
**Pull request by vlad-lesin**
TDB-27 : Add ability to kill query that is awaiting FT locktree lock
and
TDB-26 : Implement ability to use MySQL DEBUG_SYNC facility within PerconaFT

1) The ability to kill query blocked in lock-tree is added, see the initial pull request https://github.com/percona/PerconaFT/pull/360 .
2) The ability to use mysql debug_sync for mtr testing is added

See also:
https://github.com/percona/percona-server/pull/1738
https://github.com/percona/percona-server/pull/1739

MTR testing:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1896/
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/934/

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,15 @@ project(TokuDB)
 set(CMAKE_SHARED_LIBRARY_LINK_C_FLAGS "")
 set(CMAKE_SHARED_LIBRARY_LINK_CXX_FLAGS "")
 
+# detect when we are being built as a subproject
+if (DEFINED MYSQL_PROJECT_NAME_DOCSTRING)
+  add_definitions( -DMYSQL_TOKUDB_ENGINE=1)
+  if (ENABLE_DEBUG_SYNC)
```
Comment (georgelorchpercona) [inline CMakeLists.txt]: Should this be ENABLED_DEBUG_SYNC?

**Comment by georgelorchpercona** [comment]
This does not build for me with an out of source Percona Server 5.7 build using the following cmake incantation :
`cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../percona-server-install-5.7 -DMYSQL_MAINTAINER_MODE=OFF -DWITH_EMBEDDED_SERVER=OFF -DWITH_JEMALLOC=system -DWITH_BOOST=../../boost_1_59_0 -DZLIB_INCLUDE_DIR=/usr/include/ $@ ../percona-server && make -j8 -k`

The failure is :
`[ 27%] Building CXX object storage/tokudb/PerconaFT/portability/CMakeFiles/tokuportability_static_conv.dir/file.cc.o
In file included from /ssd/percona/TDB-2/percona-server/storage/tokudb/PerconaFT/portability/toku_portability.h:124:0,
                 from /ssd/percona/TDB-2/percona-server/storage/tokudb/PerconaFT/portability/file.cc:39:
/ssd/percona/TDB-2/percona-server/storage/tokudb/PerconaFT/portability/toku_debug_sync.h:45:22: fatal error: m_string.h: No such file or directory
 #include "m_string.h"
                      ^
compilation terminated.`

So I believe the s/ENABLE_DEBUG_SYNC/ENABLED_DEBUG_SYNC in CMakeLists.txt is part of the problem, but it still does not seem to be defined/passed to PerconaFT at cmake time, yet compile time is seeing it defined.

File: CMakeLists.txt (hunk 1 above)
Comment (vlad-lesin) [inline CMakeLists.txt]: No, this two definitions are different in cmake files (ENABLE_DEBUG_SYNC) and C++-source code (ENABLED_DEBUG_SYNC). See CMakeLists.txt in PS root:

```
OPTION(ENABLE_DEBUG_SYNC "Enable debug sync (debug builds only)" ON) 
IF(ENABLE_DEBUG_SYNC) 
  SET(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC") 
  SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC") 
ENDIF() 
```

**Comment by vlad-lesin** [comment]
Sorry, George, but I can't repeat the issue with cmake options you described. I also tried "release" build and it worked well. And, besides, jenkins testing did not catch it too. But I usually wipe out build directory. Maybe the issue is your previous builds influence somehow on your current build, could you try to build it in empty directory?

**Comment by georgelorchpercona** [comment]
OK, it builds if the server side is your branch from here that is the PR https://github.com/vlad-lesin/percona-server/tree/tokudb-kill-query-in-locktree-5.7-gca but not with just the 5.7 trunk alone.

Doing a diff of your PS gca branch against 5.7 HEAD right away I see :
`diff --git a/CMakeLists.txt b/CMakeLists.txt
...
-# Always enable debug sync for debug builds.
-SET(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC")
-SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC")
+OPTION(ENABLE_DEBUG_SYNC "Enable debug sync (debug builds only)" ON) 
+IF(ENABLE_DEBUG_SYNC) 
+  SET(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC") 
+  SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DENABLED_DEBUG_SYNC") 
+ENDIF() 
...`

This change is 8 years old. So then looking at your branch I see "This branch is 2 commits ahead, 823 commits behind percona:5.7." :O

So something is wrong here and this change is not compatible with 5.7 as-is.

**Comment by vlad-lesin** [comment]
George, I believe I have fixed it.  Do you still have build issues?

**Pull request merged**